### ISSUE TITLE
ERROR:  column "dbcrud_uuid" of relation "activity" does not exist

### ISSUE THREAD

**Issue by frdmn** [znuny/OTRSCloneDB#3]
While trying to use OTRSCloneDB to convert our current MySQL (8.0) installation to a PostgreSQL one, we ran into the following error:

```
znuny@ticket:~$ ./bin/znuny.Console.pl Maint::Database::Clone
Meaningful start message...
[2024-08-28 11:04:58] Generating DDL for Znuny.
[2024-08-28 11:04:58] Generating DDL for package CloneDB.
[2024-08-28 11:04:58] Generating DDL for package Znuny-ExtraSkins.
[2024-08-28 11:04:58] Creating structures in target database (phase 1/2)................................................................................................................................................................................................................................................................. done.
[2024-08-28 11:05:01] Converting table acl...
[2024-08-28 11:05:01] Finished converting table acl.
[2024-08-28 11:05:01] Converting table acl_sync...
[2024-08-28 11:05:01] Finished converting table acl_sync.
[2024-08-28 11:05:01] Converting table acl_ticket_attribute_relations...
[2024-08-28 11:05:01] Finished converting table acl_ticket_attribute_relations.
[2024-08-28 11:05:01] Converting table activity...
[Wed Aug 28 11:05:02 2024] znuny.Console.pl: DBD::Pg::db do failed: ERROR:  column "dbcrud_uuid" of relation "activity" does not exist
[Wed Aug 28 11:05:02 2024] znuny.Console.pl: LINE 1: ...ity_state, activity_link, create_time, create_by, dbcrud_uui...
[Wed Aug 28 11:05:02 2024] znuny.Console.pl:                                                              ^ at /opt/znuny-7.0.15/Kernel/System/DB.pm line 497.
ERROR: Znuny-znuny.Console.pl-Maint::Database::Clone-10 Perl: 5.30.0 OS: linux Time: Wed Aug 28 11:05:02 2024

 Message: ERROR:  column "dbcrud_uuid" of relation "activity" does not exist
LINE 1: ...ity_state, activity_link, create_time, create_by, dbcrud_uui...
                                                             ^, SQL: 'INSERT INTO activity (id, user_id, activity_type, activity_title, activity_text, activity_state, activity_link, create_time, create_by, dbcrud_uuid) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'

 Traceback (6594):
   Module: Kernel::System::CloneDB::Driver::Base::DataTransfer Line: 383
   Module: Kernel::System::CloneDB::Backend::DataTransfer Line: 228
   Module: Kernel::System::Console::Command::Maint::Database::Clone::Run Line: 108
   Module: (eval) Line: 461
   Module: Kernel::System::Console::BaseCommand::Execute Line: 455
   Module: Kernel::System::Console::InterfaceConsole::Run Line: 81
   Module: ./bin/znuny.Console.pl Line: 37

Could not insert data: Table: activity - id:4.
Error: Died at /opt/znuny-7.0.15/Kernel/System/CloneDB/Driver/Base.pm line 389.
```

The corresponding structure of the `activity` table in MySQL looks currently like this:

```mysql
mysql> DESCRIBE activity;
+----------------+--------------+------+-----+---------+----------------+
| Field          | Type         | Null | Key | Default | Extra          |
+----------------+--------------+------+-----+---------+----------------+
| id             | int          | NO   | PRI | NULL    | auto_increment |
| user_id        | int          | NO   | MUL | NULL    |                |
| activity_type  | varchar(200) | NO   |     | NULL    |                |
| activity_title | varchar(255) | NO   |     | NULL    |                |
| activity_text  | longblob     | YES  |     | NULL    |                |
| activity_state | varchar(255) | YES  |     | NULL    |                |
| activity_link  | varchar(255) | YES  |     | NULL    |                |
| create_time    | datetime     | NO   |     | NULL    |                |
| create_by      | int          | NO   | MUL | NULL    |                |
| dbcrud_uuid    | varchar(36)  | YES  | UNI | NULL    |                |
+----------------+--------------+------+-----+---------+----------------+
10 rows in set (0,00 sec)
```

In the Postgres it looks like this:

```postgres
template1=# \d activity
                                            Table "public.activity"
     Column     |              Type              | Collation | Nullable |               Default
----------------+--------------------------------+-----------+----------+--------------------------------------
 id             | integer                        |           | not null | nextval('activity_id_seq'::regclass)
 user_id        | integer                        |           | not null |
 activity_type  | character varying(200)         |           | not null |
 activity_title | character varying(255)         |           | not null |
 activity_text  | text                           |           |          |
 activity_state | character varying(255)         |           |          |
 activity_link  | character varying(255)         |           |          |
 create_time    | timestamp(0) without time zone |           | not null |
 create_by      | integer                        |           | not null |
Indexes:
    "activity_pkey" PRIMARY KEY, btree (id)
```

It looks like the `dbcrud_uuid` was not created in the PostgreSQL schema. I also tried with a fresh 7.0.15 installation without existing data, but the error message was the same upon converting.

Any idea how to fix this?

**Comment by dennykorsukewitz** [comment]
Hi @frdmn 

The column should actually be created on its own.

You can use this command as a workaround for now.

```
bin/znuny.Console.pl Admin::DBCRUD::AddUUIDColumns
```
I will check the issue further later.

best regards 
Denny

---

Internal Issue: 14

**Comment by frdmn** [comment]
Hi @dennykorsukewitz

I tried the `Admin::DBCRUD::AddUUIDColumns` command, but it seems like it does that on the current active DB, where these columns/values already exist:

```
znuny@ticket:~$ bin/znuny.Console.pl Admin::DBCRUD::AddUUIDColumns
Checking/creating UUID columns for DBCRUD tables...

Activity: Column dbcrud_uuid in table activity...
	Column already exists.
Calendar::Plugin: Column dbcrud_uuid in table calendar_appointment_plugin...
	Column already exists.
OAuth2Token: Column dbcrud_uuid in table oauth2_token...
	Column already exists.
OAuth2TokenConfig: Column dbcrud_uuid in table oauth2_token_config...
	Column already exists.

Done.
```

The problem seems to be that they are not migrated over to the destination DB when using the OTRSCloneDB plugin.

At least the error message mentioned in the original issue still occur after running the `Maint::Database::Clone` again.

Thanks for looking into this!

**Comment by dennykorsukewitz** [comment]
Hi @frdmn ,

OTRSCloneDB uses the schema.xml to create the new database.

In the current  [schema.xml](https://github.com/znuny/Znuny/blob/dev/scripts/database/schema.xml#L41) the column `dbcrud_uuid` is missing for table "activity".
Therefore it is not created.

However, it already exists in your source database, probably because you have already migrated the framework and it will then be created automatically.

I will include the column in the schema.xml of the framework 7.1.x.

---

`7.0 is not supported anymore.`

As a workaround I can recommend this here at the target framework to update the schema.xml and co.

1. Add schema.xml into your target system (scripts/database/schema.xml): 

[schema.xml.zip](https://github.com/user-attachments/files/16805562/schema.xml.zip)


   I have to zip that file, because github...:
```
We don’t support that file type.

Try again with GIF, JPEG, JPG, MOV, MP4, PNG, SVG, WEBM, CPUPROFILE, CSV, DMP, DOCX, FODG, FODP, FODS, FODT, GZ, JSON, JSONC, LOG, MD, ODF, ODG, ODP, ODS, ODT, PATCH, PDF, PPTX, TGZ, TXT, XLS, XLSX or ZIP.
```

2. Run commands in target framework:

```bash
cd /opt/znuny
FRAMEWORK=znuny
SCHEMA_FILE=$(find scripts/database -type f -name '*schema.xml')
SCHEMA_FILE_NAME=$(find scripts/database -type f -name '*schema.xml' | cut -d/ -f3- | sed 's/\.xml$//')
bin/$FRAMEWORK.Console.pl Dev::Tools::Database::XML2SQL --database-type=all --source-path=$SCHEMA_FILE --target-directory=scripts/database --target-filename=$SCHEMA_FILE_NAME --split-files
```

Denny 🚀

**Comment by dennykorsukewitz** [comment]
Fixed in Framework for 7.1.3
https://github.com/znuny/Znuny/commit/72d38b7a4d7a9582e3e6dfd4ae06cb2942f607dc
